File: MiniDB_Projects/Team_B+inC++/src/query/lexer.cpp
#include "lexer.hpp"

#include <cctype>
#include <stdexcept>
#include <unordered_map>

namespace {
const std::unordered_map<std::string, Tok>& keywords() {
    static const std::unordered_map<std::string, Tok> kw = {
        {"SELECT", Tok::SELECT}, {"FROM", Tok::FROM}, {"WHERE", Tok::WHERE},
        {"INSERT", Tok::INSERT}, {"INTO", Tok::INTO}, {"VALUES", Tok::VALUES},
        {"DELETE", Tok::DELETE}, {"CREATE", Tok::CREATE}, {"TABLE", Tok::TABLE},
        {"JOIN", Tok::JOIN}, {"ON", Tok::ON}, {"AND", Tok::AND}, {"OR", Tok::OR},
        {"INT", Tok::INT_KW}, {"TEXT", Tok::TEXT_KW},
        {"PRIMARY", Tok::PRIMARY}, {"KEY", Tok::KEY},
        {"BEGIN", Tok::BEGIN}, {"COMMIT", Tok::COMMIT}, {"ROLLBACK", Tok::ROLLBACK},
    };
    return kw;
}

std::string upper(std::string s) {
    for (char& c : s) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
    return s;
}
}  // namespace

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> out;
    const std::size_t n = sql_.size();

    while (pos_ < n) {
        char c = sql_[pos_];

        if (std::isspace(static_cast<unsigned char>(c))) { ++pos_; continue; }

        // ident or keyword
        if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            std::size_t start = pos_;
            while (pos_ < n && (std::isalnum(static_cast<unsigned char>(sql_[pos_])) || sql_[pos_] == '_'))
                ++pos_;
            std::string word = sql_.substr(start, pos_ - start);
            auto it = keywords().find(upper(word));
            if (it != keywords().end()) out.push_back({it->second, word});
            else                        out.push_back({Tok::IDENT, word});
            continue;
        }

        // number
        if (std::isdigit(static_cast<unsigned char>(c))) {
            std::size_t start = pos_;
            while (pos_ < n && std::isdigit(static_cast<unsigned char>(sql_[pos_]))) ++pos_;
            out.push_back({Tok::NUMBER, sql_.substr(start, pos_ - start)});
            continue;
        }

        // string literal
        if (c == '\'') {
            ++pos_;
            std::size_t start = pos_;
            while (pos_ < n && sql_[pos_] != '\'') ++pos_;
            if (pos_ >= n) throw std::runtime_error("unterminated string literal");
            std::string s = sql_.substr(start, pos_ - start);
            ++pos_;
            out.push_back({Tok::STRING, s});
            continue;
        }

        // operators / punctuation
        switch (c) {
            case '(': out.push_back({Tok::LPAREN, "("}); ++pos_; break;
            case ')': out.push_back({Tok::RPAREN, ")"}); ++pos_; break;
            case ',': out.push_back({Tok::COMMA,  ","}); ++pos_; break;
            case ';': out.push_back({Tok::SEMI,   ";"}); ++pos_; break;
            case '.': out.push_back({Tok::DOT,    "."}); ++pos_; break;
            case '*': out.push_back({Tok::STAR,   "*"}); ++pos_; break;
            case '=': out.push_back({Tok::EQ,     "="}); ++pos_; break;
            case '<':
                if (pos_ + 1 < n && sql_[pos_ + 1] == '=') { out.push_back({Tok::LE, "<="}); pos_ += 2; }
                else { out.push_back({Tok::LT, "<"}); ++pos_; }
                break;
            case '>':
                if (pos_ + 1 < n && sql_[pos_ + 1] == '=') { out.push_back({Tok::GE, ">="}); pos_ += 2; }
                else { out.push_back({Tok::GT, ">"}); ++pos_; }
                break;
            case '!':
                if (pos_ + 1 < n && sql_[pos_ + 1] == '=') { out.push_back({Tok::NEQ, "!="}); pos_ += 2; }
                else throw std::runtime_error("unexpected '!' (did you mean '!='?)");
                break;
            default:
                throw std::runtime_error(std::string("unexpected character: ") + c);
        }
    }

    out.push_back({Tok::END, ""});
    return out;
}

File: MiniDB_Projects/Team_XO/internal/sql/lexer.go
package sql

import (
	"fmt"
	"strings"
)

type tokenKind int

const (
	tEOF tokenKind = iota
	tIdent
	tNumber
	tString
	tPunct // ( ) , . ;
	tOp    // = != <> < <= > >=
	tStar  // *
	tKeyword
)

type token struct {
	kind tokenKind
	text string
	pos  int
}

// keywords recognised by the lexer. They are matched case-insensitively and
// normalised to upper case so the parser can compare against canonical forms.
var keywords = map[string]bool{
	"CREATE": true, "TABLE": true, "INDEX": true, "ON": true, "PRIMARY": true, "KEY": true,
	"INSERT": true, "INTO": true, "VALUES": true,
	"SELECT": true, "FROM": true, "WHERE": true, "JOIN": true, "INNER": true, "AND": true,
	"DELETE": true, "EXPLAIN": true, "COUNT": true, "AS": true,
	"BEGIN": true, "COMMIT": true, "ROLLBACK": true, "ABORT": true,
	"INT": true, "INTEGER": true, "TEXT": true, "BOOL": true, "BOOLEAN": true,
	"TRUE": true, "FALSE": true,
}

// lexer scans SQL source into a token slice. Errors are reported lazily through
// the token stream by returning an error from tokenize.
type lexer struct {
	src string
	pos int
}

func tokenize(src string) ([]token, error) {
	l := &lexer{src: src}
	var toks []token
	for {
		tok, err := l.next()
		if err != nil {
			return nil, err
		}
		toks = append(toks, tok)
		if tok.kind == tEOF {
			return toks, nil
		}
	}
}

func (l *lexer) next() (token, error) {
	l.skipSpaceAndComments()
	if l.pos >= len(l.src) {
		return token{kind: tEOF, pos: l.pos}, nil
	}
	start := l.pos
	c := l.src[l.pos]

	switch {
	case c == '\'':
		return l.lexString()
	case isDigit(c) || (c == '-' && l.pos+1 < len(l.src) && isDigit(l.src[l.pos+1])):
		return l.lexNumber()
	case isIdentStart(c):
		return l.lexIdent()
	}

	switch c {
	case '(', ')', ',', '.', ';':
		l.pos++
		return token{kind: tPunct, text: string(c), pos: start}, nil
	case '*':
		l.pos++
		return token{kind: tStar, text: "*", pos: start}, nil
	case '=':
		l.pos++
		return token{kind: tOp, text: "=", pos: start}, nil
	case '!':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tOp, text: "!=", pos: start}, nil
		}
	case '<':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tOp, text: "<=", pos: start}, nil
		}
		if l.peekAt(1) == '>' {
			l.pos += 2
			return token{kind: tOp, text: "!=", pos: start}, nil
		}
		l.pos++
		return token{kind: tOp, text: "<", pos: start}, nil
	case '>':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tOp, text: ">=", pos: start}, nil
		}
		l.pos++
		return token{kind: tOp, text: ">", pos: start}, nil
	}
	return token{}, fmt.Errorf("sql: unexpected character %q at position %d", c, l.pos)
}

func (l *lexer) lexString() (token, error) {
	start := l.pos
	l.pos++ // opening quote
	var sb strings.Builder
	for l.pos < len(l.src) {
		c := l.src[l.pos]
		if c == '\'' {
			// Doubled quote is an escaped single quote.
			if l.peekAt(1) == '\'' {
				sb.WriteByte('\'')
				l.pos += 2
				continue
			}
			l.pos++
			return token{kind: tString, text: sb.String(), pos: start}, nil
		}
		sb.WriteByte(c)
		l.pos++
	}
	return token{}, fmt.Errorf("sql: unterminated string literal starting at %d", start)
}

func (l *lexer) lexNumber() (token, error) {
	start := l.pos
	if l.src[l.pos] == '-' {
		l.pos++
	}
	for l.pos < len(l.src) && isDigit(l.src[l.pos]) {
		l.pos++
	}
	return token{kind: tNumber, text: l.src[start:l.pos], pos: start}, nil
}

func (l *lexer) lexIdent() (token, error) {
	start := l.pos
	for l.pos < len(l.src) && isIdentPart(l.src[l.pos]) {
		l.pos++
	}
	text := l.src[start:l.pos]
	upper := strings.ToUpper(text)
	if keywords[upper] {
		return token{kind: tKeyword, text: upper, pos: start}, nil
	}
	return token{kind: tIdent, text: text, pos: start}, nil
}

func (l *lexer) skipSpaceAndComments() {
	for l.pos < len(l.src) {
		c := l.src[l.pos]
		if c == ' ' || c == '\t' || c == '\n' || c == '\r' {
			l.pos++
			continue
		}
		// SQL line comments: -- to end of line.
		if c == '-' && l.peekAt(1) == '-' {
			for l.pos < len(l.src) && l.src[l.pos] != '\n' {
				l.pos++
			}
			continue
		}
		return
	}
}

func (l *lexer) peekAt(offset int) byte {
	if l.pos+offset < len(l.src) {
		return l.src[l.pos+offset]
	}
	return 0
}

func isDigit(c byte) bool      { return c >= '0' && c <= '9' }
func isIdentStart(c byte) bool { return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') }
func isIdentPart(c byte) bool  { return isIdentStart(c) || isDigit(c) }
